config/v1/types.go

```diff
@@ -5,6 +5,84 @@ import (
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Format is the format of the name to be built for the system component
+	Format string `json:"format"`
+	// Latest determines if the latest tag will be pulled from the registry
+	Latest bool `json:"latest"`
+
+	// MaxImagesBulkImportedPerRepository controls the number of images that are imported when a user
+	// does a bulk import of a Docker repository. This number defaults to 50 to prevent users from
+	// importing large numbers of images accidentally. Set -1 for no limit.
+	MaxImagesBulkImportedPerRepository int32 `json:"maxImagesBulkImportedPerRepository"`
+	// DisableScheduledImport allows scheduled background import of images to be disabled.
+	DisableScheduledImport bool `json:"disableScheduledImport"`
+	// ScheduledImageImportMinimumIntervalSeconds is the minimum number of seconds that can elapse between when image streams
+	// scheduled for background import are checked against the upstream repository. The default value is 15 minutes.
+	ScheduledImageImportMinimumIntervalSeconds int32 `json:"scheduledImageImportMinimumIntervalSeconds"`
+	// MaxScheduledImageImportsPerMinute is the maximum number of scheduled image streams that will be imported in the
+	// background per minute. The default value is 60. Set to -1 for unlimited.
+	MaxScheduledImageImportsPerMinute int32 `json:"maxScheduledImageImportsPerMinute"`
+	// AllowedRegistriesForImport limits the docker registries that normal users may import
+	// images from. Set this list to the registries that you trust to contain valid Docker
+	// images and that you want applications to be able to import from. Users with
+	// permission to create Images or ImageStreamMappings via the API are not affected by
+	// this policy - typically only administrators or system integrations will have those
+	// permissions.
+	AllowedRegistriesForImport []RegistryLocation `json:"allowedRegistriesForImport,omitempty"`
+	// InternalRegistryHostname sets the hostname for the default internal image
+	// registry. The value must be in "hostname[:port]" format.
+	// For backward compatibility, users can still use OPENSHIFT_DEFAULT_REGISTRY
+	// environment variable but this setting overrides the environment variable.
+	InternalRegistryHostname string `json:"internalRegistryHostname,omitempty"`
+	// ExternalRegistryHostname sets the hostname for the default external image
+	// registry. The external hostname should be set only when the image registry
+	// is exposed externally. The value is used in 'publicDockerImageRepository'
+	// field in ImageStreams. The value must be in "hostname[:port]" format.
+	ExternalRegistryHostname string `json:"externalRegistryHostname,omitempty"`
+	// AdditionalTrustedCA is a path to a pem bundle file containing additional CAs that
+	// should be trusted during imagestream import.
+	AdditionalTrustedCA ConfigMapReference `json:"additionalTrustedCA,omitempty"`
```

Comment:
I think this needs to be common and I made a struct that gives an alternative to the old filename approach.

Reply:
i've got no problem w/ making it a configmapreference, but since the old code read the CAs from the path defined here, this means the code needs to be updated to read the configmap instead.

Furthermore is the expectation that code that cares about the CAs should setup a watch on the configmap?  or otherwise cache the value?  Or we're ok with every time we go through a codepath that needs the additional CAs, we make an api call to fetch the configmap?

Reply:
> i've got no problem w/ making it a configmapreference, but since the old code read the CAs from the path defined here, this means the code needs to be updated to read the configmap instead.

I don't think that's the case.  I think that your operator would take care of plumbing the reference here down into a mountable configmap local to your pod and you could simply mount that new configmap and reference it. That avoid granting exposed processes powers to read configmaps in other namespaces and keeps API code out of binaries that don't need it.  Also, it makes code more easily testable because you don't need an apiserver up in order to read a config value.

Reply:
what operator?  We don't have an "image import api" operator, so is the openshift-api operator going to do it?

Reply:
and how does the code in my pod know where to read the value from?  it used to get it from the filepath stored in "AdditionalTrustedCA" that it found in the master-config, but if that's a configmap reference that's no longer the way it can be done.

Reply:
> what operator? We don't have an "image import api" operator, so is the openshift-api operator going to do it?

Sadly, yes.  By organizing by feature, we have a config driven by the user goal, not the topology of the cluster.



> and how does the code in my pod know where to read the value from? it used to get it from the filepath stored in "AdditionalTrustedCA" that it found in the master-config, but if that's a configmap reference that's no longer the way it can be done.

We have examples of how to do this inside of the openshift-apiserver operator where you can copy configmaps and secrets and then reference them in your config: https://github.com/openshift/cluster-openshift-apiserver-operator/blob/master/pkg/operator/sync_openshiftapiserver_v311_00.go#L103-L116

The actual binary never knows the difference because it is driven by a config file on disk to enable unit testing.